### ISSUE TITLE
No means to specify group membership for client SSL cert authn

### ISSUE THREAD

**Issue by treed** [kubernetes/kubernetes#30260]
I guess this is more of a feature request than a bug report, but:

With 1.3 out, I decided to look into using RBAC, but I found that a pretty significant shortcoming is that I can't specify group membership when using client certs for authn. At least, I'm pretty sure I can't. I just spent a bit looking through the code, as well.

If the username comes from the CN, then I'd expect the OUs to define group membership, but that seems not to be the case.

**Comment by treed** [comment]
To elaborate, the wall I hit here was that the kubelets aren't capable of authenticating with the kubelet with RBAC turned on, unless I add rules describing the full list of nodes, which would need to be updated every time the set of nodes changes. Each node has its own TLS cert with its own hostname as the CN.

I guess I could just issue a common "node" CN to all the nodes, but it seems like this is functionality that should exist, so that I can just set OU=node on these certs and then describe permissions that way. Would also be nice for managing users.

**Comment by AdoHe** [comment]
@kubernetes/kubectl

**Comment by zhouhaibing089** [comment]
@gtank if I understand correctly, this is the use case for kubelet cert authentication, once the tls kubelet bootstrap is done, we need to consider the RBAC policy part as well.

cc @ericchiang

**Comment by zhouhaibing089** [comment]
I would also recommend to add the `area/auth` label here..

**Comment by smarterclayton** [comment]
@kubernetes/sig-auth

**Comment by smarterclayton** [comment]
Moving to the auth area, not sig api machinery.

**Comment by ericchiang** [comment]
> I guess I could just issue a common "node" CN to all the nodes, but it seems like this is functionality that should exist, so that I can just set OU=node on these certs and then describe permissions that way. Would also be nice for managing users.

This would be easy to add, seems reasonable, and address some issues being discussed in #30094. If other @kubernetes/sig-auth members are okay with this I'd be happy to send a PR.

**Comment by deads2k** [comment]
> This would be easy to add, seems reasonable, and address some issues being discussed in #30094. If other @kubernetes/sig-auth members are okay with this I'd be happy to send a PR.

We've done this in openshift using the "Organizations" part of the subject.  There are upsides and downsides.  Let's add this to the sig-auth discussion tomorrow.  The interactions, particularly with a groupification API (coreos still want that?), are non-obvious.  We found that people definitely wanted X and definitely wanted NotX.

**Comment by treed** [comment]
FWIW, I tested a one line change to do this last night (although my hyperkube build seems to have a broken cAdvisor? Had to use stock 1.3.4 for the kubelet side).

It appears to work (in that I can get errors about disallowed actions to go away by adding permissions to the ClusterRoleBinding), but it's still having trouble reporting itself as Ready, which seems maybe related to this error (there are a few variants with different Reasons):

> Aug 09 10:18:43 AF002266 docker[20445]: E0809 10:18:43.766717   20481 event.go:198] Server rejected event '&api.Event{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"af002266.14691c0663e3c691", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil)}, InvolvedObject:api.ObjectReference{Kind:"Node", Namespace:"", Name:"af002266", UID:"af002266", APIVersion:"", ResourceVersion:"", FieldPath:""}, Reason:"NodeHasSufficientDisk", Message:"Node af002266 status is now: NodeHasSufficientDisk", Source:api.EventSource{Component:"kubelet", Host:"af002266"}, FirstTimestamp:unversioned.Time{Time:time.Time{sec:63606334367, nsec:84430993, loc:(*time.Location)(0x5887500)}}, LastTimestamp:unversioned.Time{Time:time.Time{sec:63606334723, nsec:756743212, loc:(*time.Location)(0x5887500)}}, Count:55, Type:"Normal"}': 'the server does not allow access to the requested resource (post events)' (will not retry!)

I've added "post" and "events" to the CR:

```
{
    "kind": "ClusterRole",
    "apiVersion": "rbac.authorization.k8s.io/v1alpha1",
    "metadata": {
        "name": "kubelet"
    },
    "rules": [
        {
            "apiGroups": [""],
            "resources": ["pods", "services"],
            "verbs": ["get", "list", "watch"],
            "nonResourceURLs": []
        },
        {
            "apiGroups": [""],
            "resources": ["secrets"],
            "verbs": ["get", "list"],
            "nonResourceURLs": []
        },
        {
            "apiGroups": [""],
            "resources": ["nodes"],
            "verbs": ["get", "list", "watch", "patch", "post"],
            "nonResourceURLs": []
        },
        {
            "apiGroups": [""],
            "resources": ["events"],
            "verbs": ["patch", "post"],
            "nonResourceURLs": []
        }
    ]
}
```

But this is the one such error that I can't make go away.

I guess at this point it might be a separate issue, but documenting the permissions needed for a kubelet to do its job would be very helpful.

**Comment by liggitt** [comment]
the verbs are API verbs, so it needs "get","list","create","update","patch" verbs on events

**Comment by liggitt** [comment]
and it needs access to the subresources for pods and nodes "pods/status" and "nodes/status"

**Comment by treed** [comment]
Thanks, I added those and the events errors went away.

Now the only remaining errors are about inability to cleanup orphaned pods/pod directories, which has been around for a while and I think is related to another issue (which I can't find at the moment).

However, the node is still not showing up as Ready. Unsure what else could be missing.

**Comment by deads2k** [comment]
> Thanks, I added those and the events errors went away.
> 
> Now the only remaining errors are about inability to cleanup orphaned pods/pod directories, which has been around for a while and I think is related to another issue (which I can't find at the moment).
> 
> However, the node is still not showing up as Ready. Unsure what else could be missing.

Take a look at openshift's node role: https://github.com/openshift/origin/blob/master/pkg/cmd/server/bootstrappolicy/policy.go#L558-L599  We've been using since 1.0, so it should have everything you need.

**Comment by deads2k** [comment]
Oh, they're bound to this too: https://github.com/openshift/origin/blob/master/pkg/cmd/server/bootstrappolicy/policy.go#L522-L530 .  There are other node related roles in there if you wish start wiring up a node authorizer to protect your node.

**Comment by erictune** [comment]
I expect group memberships can be changed pretty quickly and propagate
quickly.
Our certs last a long time.
So putting a group membership in a cert, while convenient for the special
case of a kubelet, doesn't seem like a pattern to encourage elsewhere.  A
source of user to group membership mappings, which is independent of
authentication method, seems like it is what is needed.  And, of course,
there is already an issue for that:
https://github.com/kubernetes/kubernetes/issues/23720

**Comment by treed** [comment]
Thanks for the quick turnaround on this!
